### PR TITLE
Replace ScalarQueryUtil with QueryCardinalityUtil

### PR DESCRIPTION
**Pull request by kokosing**
Replace ScalarQueryUtil with QueryCardinalityUtil

ScalarQueryUtil was wrong. Its goal semantic was to tell if given query
is scalar (always return a row), but it was returning true for queries
wrapped with FilterNode for which it is not known how many rows will be
returned.

QueryCardinalityUtil solves that as it returns a cardinality range,
where for scalar queries range is <1,1>.

### CODE REVIEW

**Comment by kokosing** [comment]
It refers to issue raised by master failures after https://github.com/prestodb/presto/pull/8312 (fixed by https://github.com/prestodb/presto/pull/8333)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/QueryCardinalityUtil.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        public Range<Long> visitValues(ValuesNode node, Void context)
+        {
+            return Range.singleton((long) node.getRows().size());
+        }
+
+        @Override
+        public Range<Long> visitLimit(LimitNode node, Void context)
+        {
+            Range<Long> sourceCardinalityRange = node.getSource().accept(this, null);
+            return Range.closed(sourceCardinalityRange.lowerEndpoint(), min(sourceCardinalityRange.upperEndpoint(), node.getCount()));
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/QueryCardinalityUtil.java]: make sure lowerEndpoint is lower than higherEndpoint

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/QueryCardinalityUtil.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+import static java.lang.Math.min;
+import static java.util.Objects.requireNonNull;
+
+public final class QueryCardinalityUtil
+{
+    private QueryCardinalityUtil()
+    {
+    }
+
+    public static boolean isScalar(PlanNode node)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/QueryCardinalityUtil.java]: maybe add base method `isAtMostCardinality(PlanNode node, long cardinality, Lookup lookup)`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/QueryCardinalityUtil.java (hunk 1 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/QueryCardinalityUtil.java]: `sourceCardinalityRange.lowerEndpoint()` might higher than `node.getCount()`, so range will be incorrect

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/QueryCardinalityUtil.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+        @Override
+        public Range<Long> visitLimit(LimitNode node, Void context)
+        {
+            Range<Long> sourceCardinalityRange = node.getSource().accept(this, null);
+            return Range.closed(sourceCardinalityRange.lowerEndpoint(), min(sourceCardinalityRange.upperEndpoint(), node.getCount()));
+        }
+    }
+
+    public static void main(String[] args)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/QueryCardinalityUtil.java]: add proper tests instead

**Comment by kokosing** [comment]
@sopel39 Comments addressed, though I did not add tests

**Comment by kokosing** [comment]
@sopel39 test added as well

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/QueryCardinalityUtil.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    {
+        return Range.singleton(1L).encloses(extractCardinality(node, lookup));
+    }
+
+    public static boolean isAtMostScalar(PlanNode node)
+    {
+        return isAtMostScalar(node, noLookup());
+    }
+
+    public static boolean isAtMostScalar(PlanNode node, Lookup lookup)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/QueryCardinalityUtil.java]: "at most scalar" is weird. Scalar is a description of a function that produces exactly one value vs a table function, which produces a collection of values. I'd pick a name that's more explicit about this being related to whether it produces at most one row.

Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/QueryCardinalityUtil.java]: Discussed with @martint, he doesn't have a better name right now, so we'll merge now and revisit.

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/QueryCardinalityUtil.java]: I agree that this name is not the best one

**Pull request merged**